Fix duplication of blocks when toggling between long/short instructions

fisher-alice opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior. 
-->
This PR fixes a bug in Blockly labs (both CDO and Google) in the top instructions panel - when the 'long' instructions include  blocks, these blocks duplicate when the user toggled between 'long' and 'short' instructions. 

BEFORE UPDATE:

<img width="1680" alt="image-20230317-181443" src="https://github.com/code-dot-org/code-dot-org/assets/107423305/b52c9a0e-a4ed-4b72-89ec-ff550627a810">


In  `instructions/utils.js`, I updated the function `convertXmlToBlockly` that is called on in `MarkdownInstructions.jsx`. I added a class name to the `blockSpaceContainer` element so that any existing blockspace containers can be easily removed before they are created per XML element. 

The `convertXmlToBlockly` function is also called on in `_authored_hints.haml` and `_long_instructions` so that when levelbuilders are editing markdown in edit mode, they can view the rendered blocks. There were fixes for [long instructions](https://github.com/code-dot-org/code-dot-org/pull/52037) and [authored hints](https://github.com/code-dot-org/code-dot-org/pull/51638) to address the duplicate blocks bug which can now be removed because the update in this PR resolves this duplicate block issue more generally. 

AFTER UPDATE:


https://github.com/code-dot-org/code-dot-org/assets/107423305/3749790c-77f3-4804-82c0-0b0c38f60547


<details>

<summary>Background/Context (optional)</summary>

This bug was previously solved by [this PR](https://github.com/code-dot-org/code-dot-org/pull/28453). I think there was  a regression from [this more recent PR](https://github.com/code-dot-org/code-dot-org/pull/44270) which addressed issues in rendering instructions. In particular, [this line](https://github.com/code-dot-org/code-dot-org/blob/87eefe3c2d56d8dc19affee45e0b558c82713bd9/apps/src/templates/instructions/utils.js#L144).
[Slack discussion](https://codedotorg.slack.com/archives/C0T0PNTM3/p1679075945283839)
</details>




## Links
[jira ticket](https://codedotorg.atlassian.net/browse/SL-693)
<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story
I tested locally in both Google and CDO Blockly labs that included long and short instructions. I also checked in both Google and CDO blockly levelbuilder edit mode that there were no duplicate blocks rendered in the long instructions or the authored hints markdown.

## Follow-up
I noticed that on this level, once the hints are opened, they do not close when the browser is refreshed after this update. They do disappear when the 'Less' button is clicked. This happens when on production. Unsure if this should be fixed. Will consult with curriculum.
[jira ticket](https://codedotorg.atlassian.net/browse/SL-901)
UPDATE: I checked with @mikeharv and he confirmed that this is expected behavior.

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

bencodeorg commented on apps/src/templates/instructions/utils.js:
```diff
@@ -114,8 +114,15 @@ function removeCommentNodes(root) {
  */
 export function convertXmlToBlockly(xmlContainer, isRtl) {
-  const xmls = xmlContainer.getElementsByTagName('xml');
+  // Remove any divs and/or spans with class name 'block-space-container'.
```
Nit: could you make this comment describe why its here, rather than what it does? I think the code itself describes pretty clearly what it does, but the why is less so.

It took me a minute to realize that the `block-space-container` class is only applied in this file, not a general Blockly thing (is that right?)

bencodeorg replied:
Nit: might be able to shorten this up to something like:

```
xmlContainer
  .getElementsByClassName('block-space-container)
  .forEach(container => container.remove())
```

mikeharv replied:
Agreed. I like the more concise approach of the `forEach` loop here.

mikeharv replied:
I decided to try @bencodeorg's suggestion above real quick and found that it didn't _quite_ work because [`getElementsByClassName`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByClassName) returns an array-like HTMLCollection value that doesn't have a `forEach` method. 

Fortunately, it looks like it's pretty easy to convert it to an array with either `Array.from` or the spread operator. https://dev.to/hasnaindev/easiest-way-to-convert-htmlcollection-into-an-array-bkl

I tested that both of these alternatives work. Not sure which is better, but I still think both are more readable than the standard for loop.

```
  Array.from(
    xmlContainer.getElementsByClassName('block-space-container')
  ).forEach(container => container.remove());
```

```
  [...xmlContainer.getElementsByClassName('block-space-container')].forEach(
    container => container.remove()
  );
```

Hamms replied:
`HTMLCollection` not supporting the standard Array iteration methods is such a footgun; I had the same instinct as Ben, and do pretty much every time I encounter them :upside_down_face: 

I like the `Array.from` approach, but I also think a comment that just says something like "using a for loop rather than forEach because HTMLCollection" would suffice

bencodeorg replied:
hah that is annoying! Should have known there was probably some reason you did it this way Alice :). TIL!

fisher-alice replied:
Thanks y'all! I think I also prefer the `Array.from` and will update.

fisher-alice replied:
> Nit: could you make this comment describe why its here, rather than what it does? I think the code itself describes pretty clearly what it does, but the why is less so.
> 
> It took me a minute to realize that the `block-space-container` class is only applied in this file, not a general Blockly thing (is that right?)

Yes - will add a clarifying comment here. @Hamms also suggested this. Thanks!

fisher-alice replied:
I also updated the description of the function to include where this util function is used - authored hints and long instructions in the Instructions panel.

bencodeorg reviewed (approved):
Nice!

bencodeorg commented on apps/src/templates/instructions/utils.js:
```diff
@@ -114,8 +114,15 @@ function removeCommentNodes(root) {
  * @param {Boolean} isRtl True if we are displaying in RTL
  */
 export function convertXmlToBlockly(xmlContainer, isRtl) {
```
One other comment -- I think this is ok (removing previous block rendering each time this is called), but it did raise my antenna a little bit (ie, is there a situation where this would now unexpectedly remove previously rendered blocks?) MIght be worth a) updating the description of this function to mention it, and b) checking other uses of this utility to see if its ever used in a way that wouldn't want this new behavior.

mikeharv replied:
I had the same thought, but I think I'm coming to a different conclusion. I looked through the other uses of this utility and I'm not seeing any regressions. The changes also seem like they'd make sense in those contexts; it's just that in components like `ProgrammingExpressionOverview`, we don't provide UI that causes us to "refresh" the embedded blocks.

I actually am wondering if this change could allow us to revert the previous fixes we made to `_authored_hints.haml` and `_long_instructions.html.haml` (Recent commit: https://github.com/code-dot-org/code-dot-org/pull/52034/commits/80428e103155df1b9746942045c0f0a407da84d1)

bencodeorg replied:
@mikeharv I don't think I had a different conclusion here? Your suggestions make sense to me.

mikeharv replied:
You're totally right! Apologies for misreading.

fisher-alice replied:
@mikeharv - Just checked, and this change means I can revert the fixes made in the commit you reference above and the duplicate blocks bug is fixed in `_long_instructions.html.haml` and `_authored_hints.haml`.

mikeharv commented on apps/src/templates/instructions/utils.js:
```diff
@@ -114,8 +114,15 @@ function removeCommentNodes(root) {
+  );
+  for (let i = blockSpaceContainers.length - 1; i >= 0; i--) {
+    blockSpaceContainers[i].remove();
+  }
```
When we fixed this issue in long instructions and authored hints, we went with an approach of removing the actual SVGs (with class `readOnlyBlockSpace`) rather than their containers. The SVGs created here (by `MarkdownInstructions` -> `onMainBlockSpaceCreated()`) also have this same class, so I wonder if you considered using the same approach as before here.

Either way, it seems like this issue of duplicating embedded blocks has turned into a bit of a whack-a-mole. But I'm feeling more confident that fixing it here inside the definition of `convertXmlToBlockly` is the way to go. (I initially shared Ben's reservations which is why I edited the .haml file previously.)

fisher-alice replied:
At first, I traced the call to the `convertXmlBlockly` function from the React component `MarkdownInstructions.jsx` (a component within the Instructions panel). And I just thought it would be more ~~simple~~ readable to remove the blockly container (div or span) instead of the SVGs. It's nice that we can now remove the other fixes in the .haml files.

fisher-alice replied:
Just want to document that the alternative approach works as well.
```
  const blocklySvgs = xmlContainer.querySelectorAll(
    'svg[class*="readOnlyBlockSpace"]'
  );
   blocklySvgs.forEach(svg => {
     svg.parentNode.remove();
   });
```

mikeharv commented on apps/src/templates/instructions/utils.js:
```diff
@@ -135,6 +142,7 @@ export function convertXmlToBlockly(xmlContainer, isRtl) {
 
     // create a container and insert the blockspace into it
     const blockSpaceContainer = document.createElement(inline ? 'span' : 'div');
+    blockSpaceContainer.classList.add('block-space-container');
```
I think what we're doing here is tagging these new elements so that we can find and remove them later, right? Clever!

fisher-alice replied:
Yes - that's correct. thanks!

Hamms reviewed (approved):
This approach makes sense to me! One thing I would love to see is some kind of accounting of all the other one-off fixes we've tried to apply for this issue in the past. I expect there's already some amount of redundancy here already, and with this broad fix in place there will be even more. It would be nice to remove any we don't need, and get assurance that there aren't already-dealt-with edge cases that this approach somehow misses out on. That's probably not a blocker, though, so much as an important piece of follow-up.

One change I would like to see before this gets merged is a verbose comment explaining explicitly why we're doing this and how we're creating and using the `block-space-container` class.

fisher-alice commented:
> This approach makes sense to me! One thing I would love to see is some kind of accounting of all the other one-off fixes we've tried to apply for this issue in the past. I expect there's already some amount of redundancy here already, and with this broad fix in place there will be even more. It would be nice to remove any we don't need, and get assurance that there aren't already-dealt-with edge cases that this approach somehow misses out on. That's probably not a blocker, though, so much as an important piece of follow-up.
> 
> One change I would like to see before this gets merged is a verbose comment explaining explicitly why we're doing this and how we're creating and using the `block-space-container` class.

Thanks for the suggestions! I added a detailed comment, updated the function description, and added more details in the PR description.
